### ISSUE TITLE
Wrong numbering of ucd_to_deal in GeometryInfo<3>

### ISSUE THREAD

**Issue by nlsmrg** [dealii/dealii#14608]
Right now ucd_to_deal is an array containing

> 0, 1, 5, 4, 2, 3, 7, 6

Looking at the sketch of the vertex numbering in the GeometryInfo documentation this seems correct on first sight. However, we look at it from the x-z-plane and one can easily get fooled into believing the numbering is counter-clockwise when it is in fact clockwise. 
I double checked it using the following code and the Cell Size filter in Paraview.
```
#include <deal.II/grid/grid_generator.h>
#include <deal.II/grid/grid_out.h>
#include <deal.II/grid/grid_tools.h>
#include <fstream>
int main(int argc, char **argv) {
  dealii::Utilities::MPI::MPI_InitFinalize mpi(argc, argv);
  dealii::GridOutFlags::Ucd ucd_flags(false, true, true);
  dealii::Triangulation<3> tria3;
  dealii::GridGenerator::hyper_cube(tria3);
  dealii::GridOut grid_out;
  grid_out.set_flags(ucd_flags);
  std::string filename{"test.inp"};
  std::ofstream ucd_file(filename);
  grid_out.set_flags(ucd_flags);
  grid_out.write_ucd(tria3, ucd_file);
}
```
One possible correct numbering would be for example
> 1, 0, 4, 5, 3, 2, 6, 7

**Comment by drwells** [comment]
Possibly related to #12140.

**Comment by bangerth** [comment]
@nlsmrg: I ran this program and opened the file in Paraview, but it looks like a perfectly fine cube with all edges as they should be. How does it look for you?

**Comment by nlsmrg** [comment]
@bangerth True, it does look a cube. The problem is less obvious: In the attached picture I added the CellSize Filter in Paraview. You see the volume is negative. If you read the file back into deal, I guess you won't see a problem. If you read it and want to use it in another context, you will get problems. That is the only reason I came across this: The FEM library I wanted to use warned me about the negative volumes.
![20230111_20h13m21s_grim](https://user-images.githubusercontent.com/20728434/211897721-a98fb21b-5a13-47c1-8eb5-6fcbdf586550.png)

**Comment by bangerth** [comment]
Oh, I see. They just want the whole thing inverted (or flipped across one of the midplanes). Your reordering flips it in x-direction across the y-z midplane.

I've taken a look where the `ucd_to_deal` function is used, and unfortunately it is used in quite a number of places:
```
> grep -n -r ucd_to_deal include/ source/
include/deal.II/base/geometry_info.h:48:      ucd_to_deal()
include/deal.II/base/geometry_info.h:100:      ucd_to_deal()
include/deal.II/base/geometry_info.h:158:      ucd_to_deal()
include/deal.II/base/geometry_info.h:227:      ucd_to_deal()
include/deal.II/base/geometry_info.h:1410:   *   out << cell->vertex(ucd_to_deal[i]);
include/deal.II/base/geometry_info.h:1417:  static const std::array<unsigned int, vertices_per_cell> ucd_to_deal;
include/deal.II/base/geometry_info.h:2095:   *   out << cell->vertex(ucd_to_deal[i]);
include/deal.II/base/geometry_info.h:2102:  static constexpr std::array<unsigned int, vertices_per_cell> ucd_to_deal =
include/deal.II/base/geometry_info.h:2103:    internal::GeometryInfoHelper::Initializers<dim>::ucd_to_deal();
source/base/geometry_info.cc:74:  GeometryInfo<dim>::ucd_to_deal;
source/base/geometry_info.cc:77:  GeometryInfo<0>::ucd_to_deal = {{0}};
source/base/data_out_base.cc:1642:      stream << '\t' << nodes[GeometryInfo<dim>::ucd_to_deal[i]] + 1;
source/grid/tria.cc:456:            cell.vertices[GeometryInfo<3>::ucd_to_deal[i]] = tmp[i];
source/grid/grid_generator.cc:1816:            cells[i].vertices[GeometryInfo<3>::ucd_to_deal[0 + 4 * j]] =
source/grid/grid_generator.cc:1818:            cells[i].vertices[GeometryInfo<3>::ucd_to_deal[1 + 4 * j]] =
source/grid/grid_generator.cc:1820:            cells[i].vertices[GeometryInfo<3>::ucd_to_deal[2 + 4 * j]] =
source/grid/grid_generator.cc:1822:            cells[i].vertices[GeometryInfo<3>::ucd_to_deal[3 + 4 * j]] =
source/grid/grid_generator.cc:1830:    cells[n_cells - 1].vertices[GeometryInfo<3>::ucd_to_deal[4]] =
source/grid/grid_generator.cc:1832:    cells[n_cells - 1].vertices[GeometryInfo<3>::ucd_to_deal[5]] =
source/grid/grid_generator.cc:1834:    cells[n_cells - 1].vertices[GeometryInfo<3>::ucd_to_deal[6]] =
source/grid/grid_generator.cc:1836:    cells[n_cells - 1].vertices[GeometryInfo<3>::ucd_to_deal[7]] =
source/grid/grid_reordering.cc:65:          cell.vertices[i] = tmp[GeometryInfo<3>::ucd_to_deal[i]];
source/grid/grid_reordering.cc:96:          cell.vertices[GeometryInfo<3>::ucd_to_deal[i]] = tmp[i];
source/grid/grid_in.cc:986:            in >> cells.back().vertices[GeometryInfo<dim>::ucd_to_deal[i]];
source/grid/grid_in.cc:1075:                    .vertices[GeometryInfo<2>::ucd_to_deal[i]];
source/grid/grid_in.cc:1341:          in >> cells.back().vertices[GeometryInfo<dim>::ucd_to_deal[i]];
source/grid/grid_in.cc:2480:                                .vertices[GeometryInfo<dim>::ucd_to_deal[i]];
source/grid/grid_in.cc:3324:            in >> cells[i].vertices[GeometryInfo<dim>::ucd_to_deal[j]];
source/grid/grid_in.cc:3436:                    .vertices[GeometryInfo<dim>::ucd_to_deal[f]] =
source/grid/grid_out.cc:1086:            out << cell->vertex_index(GeometryInfo<dim>::ucd_to_deal[vertex]) +
source/grid/grid_out.cc:1210:        out << cell->vertex_index(GeometryInfo<dim>::ucd_to_deal[vertex]) + 1
source/grid/grid_out.cc:1370:            cell->vertex(GeometryInfo<dim>::ucd_to_deal[k % nv]);
source/grid/grid_out.cc:3916:                       GeometryInfo<dim - 1>::ucd_to_deal[vertex]) +
source/grid/grid_out.cc:3964:                     GeometryInfo<dim - 2>::ucd_to_deal[vertex]) +
source/grid/grid_out.cc:4083:                   GeometryInfo<dim - 1>::ucd_to_deal[vertex]) +
source/grid/grid_out.cc:4123:                     GeometryInfo<dim - 2>::ucd_to_deal[vertex]) +
source/grid/grid_out.cc:4256:                out << cell->vertex(GeometryInfo<dim>::ucd_to_deal[i]) << ' '
```
Most of these are not actually related to UCD output, but are just gratuitous uses of the array for a number of other purposes. We should see which of these can be eliminated because I imagine that we'd break a lot of stuff if we just changed the underlying array.

The ones in `grid_generator.cc` should be relatively simple to eliminate by just putting the current set of numbers in place of that array. Would you like to give it a try to write a patch for that? That'd be a good starting point to address the broader issue you mention!

**Comment by nlsmrg** [comment]
Yes I'll happily write patch for that, I will get on it tomorrow.

**Comment by bangerth** [comment]
Fantastic, thank you!

**Comment by nlsmrg** [comment]
Once PR  #14681 is merged, we are almost finished. 

I think in grid_out.cc there is no pressing need to change anything. The usage of ucd_to_deal seem to be restrictred to cases where the node numberings should be compatible (gmsh and UCD numbering). The XFig and gnuplot output shouldn't be problematic: There is no implementation of write_xfig in 3D and write_gnuplot should work for an inverted version of the current ucd_to_deal array.  

In grid_in.cc it is similar: The numberings (gmsh, UCD, DB, tecplot) are compatible. I guess one of the reasons that this has not been noticed before is that every function which reads in a mesh calls apply_grid_fixup_functions. That function calls GridTools::invert_cells_with_negative_measure.

**Comment by drwells** [comment]
Fixed by #14691.
